README.md
# React-Data-Export

##### :warning: A complete re-write is coming soon (we won't need xlsx package anymore (It will be backward compatible):warning: 

## :new: I am re-writing complete excel api in JavaScript, please consider contributing or putting your ideas here https://github.com/securedeveloper/javascript-excel (After finishing will update this library)
(:exclamation::exclamation::exclamation:The purpose of having a new library is that open source libraries either does not support styling and rest functionality in excel or they are too heavy to  consider)

[![npm version](https://badge.fury.io/js/react-data-export.svg)](https://badge.fury.io/js/react-data-export)
[![dependencies Status](https://david-dm.org/securedeveloper/react-data-export/status.svg)](https://david-dm.org/securedeveloper/react-data-export)
[![devDependencies Status](https://david-dm.org/securedeveloper/react-data-export/dev-status.svg)](https://david-dm.org/securedeveloper/react-data-export?type=dev)
[![Build Status](https://travis-ci.org/securedeveloper/react-data-export.svg?branch=master)](https://travis-ci.org/securedeveloper/react-data-export)
[![Vulnerabilities](https://snyk.io/test/github/securedeveloper/react-data-export/badge.svg)](https://snyk.io/test/github/securedeveloper/react-data-export)
[![Coverage Status](https://coveralls.io/repos/github/securedeveloper/react-data-export/badge.svg?branch=master)](https://coveralls.io/github/securedeveloper/react-data-export?branch=master)


A data export library built with and for [React](http://facebook.github.io/react/index.html). 

## Installation

With [npm](https://www.npmjs.org/package/react-data-export):

```sh
npm install react-data-export --save
```

## Code Examples
* [Simple Excel Export](examples/simple_excel_export_01.md)
* [Excel Export with Dataset](examples/simple_excel_export_02.md)
* [Excel Export with Custom Download Button](examples/with_custom_download_element.md)
* [Excel Export with custom cells style](examples/styled_excel_sheet.md)

## Excel Props
| Prop          | Type                 | Default    | Required | Description                      
| :------------ | :------------------- | :--------- | :------- | :-------------------------------------------------
| hideElement	| `bool`			   | false      | `false`  | To hide the button & directly download excel file
| filename      | `string`             | Download   | `false`  | Excel file name to be downloaded 
| fileExtension | `string`             | xlsx       | `false`  | Download file extension [xlsx]
| element       | `HTMLElement`        | `<button>` | `false`  | Element to download excel file
| children      | `Array<ExcelSheet>`  |  `null`    | `true`   | ExcelSheet Represents data

### ExcelSheet Props
| Prop          | Type                    | Default    | Required | Description                      
| :------------ | :---------------------- | :--------- | :------- | :------------------------------- 
| name          | `string`                | `""`       | `true`   | Sheet name in file 
| data          | `array<object>`         | `null`     | `false`  | Excel Sheet data 
| dataSet       | `array<ExcelSheetData>` | `null`     | `false`  | Excel Sheet data
| children      | `ExcelColumn`           |  `null`    | `false`  | ExcelColumns

**Note:** In ExcelSheet props `dataSet` has `precedence` over `data` and `children` props.

For further types and definitions [Read More](types/types.md)

## Cell Style

Cell styles are specified by a style object that roughly parallels the OpenXML structure.  The style object has five
top-level attributes: `fill`, `font`, `numFmt`, `alignment`, and `border`.

| Style Attribute | Sub Attributes | Values |
| :-------------- | :------------- | :------------- |
| fill            | patternType    |  `"solid"` or `"none"`
|                 | fgColor        |  `COLOR_SPEC`
|                 | bgColor        |  `COLOR_SPEC`
| font            | name           |  `"Calibri"` // default
|                 | sz             |  `"11"` // font size in points
|                 | color          |  `COLOR_SPEC`
|                 | bold           |  `true` or `false`
|                 | underline      |  `true` or `false`
|                 | italic         |  `true` or `false`
|                 | strike         |  `true` or `false`
|                 | outline        |  `true` or `false`
|                 | shadow         |  `true` or `false`
|                 | vertAlign      |  `true` or `false`
| numFmt          |                |  `"0"`  // integer index to built in formats, see StyleBuilder.SSF property
|                 |                |  `"0.00%"` // string matching a built-in format, see StyleBuilder.SSF
|                 |                |  `"0.0%"`  // string specifying a custom format
|                 |                |  `"0.00%;\\(0.00%\\);\\-;@"` // string specifying a custom format, escaping special characters
|                 |                |  `"m/dd/yy"` // string a date format using Excel's format notation
| alignment       | vertical       | `"bottom"` or `"center"` or `"top"`
|                 | horizontal     | `"left"` or `"center"` or `"right"`
|                 | wrapText       |  `true ` or ` false`
|                 | readingOrder   |  `2` // for right-to-left
|                 | textRotation   | Number from `0` to `180` or `255` (default is `0`)
|                 |                |  `90` is rotated up 90 degrees
|                 |                |  `45` is rotated up 45 degrees
|                 |                | `135` is rotated down 45 degrees
|                 |                | `180` is rotated down 180 degrees
|                 |                | `255` is special,  aligned vertically
| border          | top            | `{ style: BORDER_STYLE, color: COLOR_SPEC }`
|                 | bottom         | `{ style: BORDER_STYLE, color: COLOR_SPEC }`
|                 | left           | `{ style: BORDER_STYLE, color: COLOR_SPEC }`
|                 | right          | `{ style: BORDER_STYLE, color: COLOR_SPEC }`
|                 | diagonal       | `{ style: BORDER_STYLE, color: COLOR_SPEC }`
|                 | diagonalUp     | `true` or `false`
|                 | diagonalDown   | `true` or `false`

**COLOR_SPEC**: Colors for `fill`, `font`, and `border` are specified as objects, either:
* `{ auto: 1}` specifying automatic values
* `{ rgb: "FFFFAA00" }` specifying a hex ARGB value
* `{ theme: "1", tint: "-0.25"}` specifying an integer index to a theme color and a tint value (default 0)
* `{ indexed: 64}` default value for `fill.bgColor`

**BORDER_STYLE**: Border style is a string value which may take on one of the following values:
 * `thin`
 * `medium`
 * `thick`
 * `dotted`
 * `hair`
 * `dashed`
 * `mediumDashed`
 * `dashDot`
 * `mediumDashDot`
 * `dashDotDot`
 * `mediumDashDotDot`
 * `slantDashDot`


Borders for merged areas are specified for each cell within the merged area.  So to apply a box border to a merged area of 3x3 cells, border styles would need to be specified for eight different cells:
* left borders for the three cells on the left,
* right borders for the cells on the right
* top borders for the cells on the top
* bottom borders for the cells on the left


## Dependencies 
This library uses [file-saver](https://www.npmjs.com/package/file-saver) and [xlsx](https://www.npmjs.com/package/xlsx) and using [json-loader](https://www.npmjs.com/package/json-loader) will do the magic for you.


```javascript
///webpack.config.js
vendor: [
        .....
        'xlsx',
        'file-saver'
],
.....
node: {fs: 'empty'},
externals: [
    {'./cptable': 'var cptable'},
    {'./jszip': 'jszip'}
 ]
```



CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.7.0"></a>
# 0.7.0 (2023-05-16)


### Bug Fixes

* Fixed[#41](https://github.com/securedeveloper/react-data-export/issues/41), Fixed[#55](https://github.com/securedeveloper/react-data-export/issues/55), Fixed[#66](https://github.com/securedeveloper/react-data-export/issues/66), Fixed[#67](https://github.com/securedeveloper/react-data-export/issues/67), Fixed[#68](https://github.com/securedeveloper/react-data-export/issues/68), Fixed[#69](https://github.com/securedeveloper/react-data-export/issues/69), Fixed[#72](https://github.com/securedeveloper/react-data-export/issues/72) ([0144afc](https://github.com/securedeveloper/react-data-export/commit/0144afc))
* improve types to represent children, fix types to represent ExcelCellData with arrays (as in ex2 - multiDataSet) ([5f0d69d](https://github.com/securedeveloper/react-data-export/commit/5f0d69d))
* int cells were not supported in dataSet format ([513ae48](https://github.com/securedeveloper/react-data-export/commit/513ae48))
* package vulnerabilities. ([082947a](https://github.com/securedeveloper/react-data-export/commit/082947a))
* package vulnerabilities. ([d66eb9e](https://github.com/securedeveloper/react-data-export/commit/d66eb9e))
* package.json to reduce vulnerabilities ([cfc5ecb](https://github.com/securedeveloper/react-data-export/commit/cfc5ecb))
* remove package.lock to stabalize ([6e66121](https://github.com/securedeveloper/react-data-export/commit/6e66121))
* updated package to tempa-xlsx, updated jest test environment and imports. ([b4f03ac](https://github.com/securedeveloper/react-data-export/commit/b4f03ac))
* **graphite:** move to npm package ([560b2f9](https://github.com/securedeveloper/react-data-export/commit/560b2f9)), closes [issue#41](https://github.com/issue/issues/41)


### Features

* add excel data export ([94530e3](https://github.com/securedeveloper/react-data-export/commit/94530e3))
* add jest support ([ff3a2ba](https://github.com/securedeveloper/react-data-export/commit/ff3a2ba))
* add react component typings ([25528dd](https://github.com/securedeveloper/react-data-export/commit/25528dd))
* if the header has style then the style should be used ([855d37a](https://github.com/securedeveloper/react-data-export/commit/855d37a))
* support col width ([dc6dc60](https://github.com/securedeveloper/react-data-export/commit/dc6dc60))
* update example with col widths ([4863aaf](https://github.com/securedeveloper/react-data-export/commit/4863aaf))



<a name="0.6.0"></a>
# [0.6.0](https://github.com/securedeveloper/react-data-export/compare/v0.5.0...v0.6.0) (2020-01-20)


### Features

* if the header has style then the style should be used ([855d37a](https://github.com/securedeveloper/react-data-export/commit/855d37a))



<a name="0.5.0"></a>
# [0.5.0](https://github.com/securedeveloper/react-data-export/compare/v0.4.2...v0.5.0) (2018-09-24)


### Bug Fixes

* Fixed[#41](https://github.com/securedeveloper/react-data-export/issues/41), Fixed[#55](https://github.com/securedeveloper/react-data-export/issues/55), Fixed[#66](https://github.com/securedeveloper/react-data-export/issues/66), Fixed[#67](https://github.com/securedeveloper/react-data-export/issues/67), Fixed[#68](https://github.com/securedeveloper/react-data-export/issues/68), Fixed[#69](https://github.com/securedeveloper/react-data-export/issues/69), Fixed[#72](https://github.com/securedeveloper/react-data-export/issues/72) ([0144afc](https://github.com/securedeveloper/react-data-export/commit/0144afc))
* int cells were not supported in dataSet format ([513ae48](https://github.com/securedeveloper/react-data-export/commit/513ae48))
* package vulnerabilities. ([082947a](https://github.com/securedeveloper/react-data-export/commit/082947a))
* package vulnerabilities. ([d66eb9e](https://github.com/securedeveloper/react-data-export/commit/d66eb9e))
* updated package to tempa-xlsx, updated jest test environment and imports. ([b4f03ac](https://github.com/securedeveloper/react-data-export/commit/b4f03ac))


### Features

* support col width ([dc6dc60](https://github.com/securedeveloper/react-data-export/commit/dc6dc60))
* update example with col widths ([4863aaf](https://github.com/securedeveloper/react-data-export/commit/4863aaf))



<a name="0.4.2"></a>
## [0.4.2](https://github.com/securedeveloper/react-data-export/compare/v0.4.1...v0.4.2) (2018-04-24)



<a name="0.4.1"></a>
## [0.4.1](https://github.com/securedeveloper/react-data-export/compare/v0.4.0...v0.4.1) (2018-04-24)



<a name="0.4.0"></a>
# [0.4.0](https://github.com/securedeveloper/react-data-export/compare/v0.3.7...v0.4.0) (2018-04-24)


### Bug Fixes

* **graphite:** move to npm package ([560b2f9](https://github.com/securedeveloper/react-data-export/commit/560b2f9)), closes [issue#41](https://github.com/issue/issues/41)


### Features

* add jest support ([ff3a2ba](https://github.com/securedeveloper/react-data-export/commit/ff3a2ba))



<a name="0.3.7"></a>
## [0.3.7](https://github.com/securedeveloper/react-data-export/compare/v0.3.6...v0.3.7) (2018-04-04)



<a name="0.3.6"></a>
## [0.3.6](https://github.com/securedeveloper/react-data-export/compare/v0.3.4...v0.3.6) (2018-04-04)



<a name="0.3.5"></a>
## [0.3.5](https://github.com/securedeveloper/react-data-export/compare/v0.3.4...v0.3.5) (2018-04-04)



<a name="0.3.4"></a>
## [0.3.4](https://github.com/securedeveloper/react-data-export/compare/v0.3.3...v0.3.4) (2018-04-04)



<a name="0.3.3"></a>
## [0.3.3](https://github.com/securedeveloper/react-data-export/compare/v0.3.0...v0.3.3) (2018-04-02)


### Bug Fixes

* package.json to reduce vulnerabilities ([cfc5ecb](https://github.com/securedeveloper/react-data-export/commit/cfc5ecb))



<a name="0.3.2"></a>

types/index.d.ts
/* index.d.ts (C) react-data-export */

// TypeScript Version: 2.2
declare module 'react-data-export' {
  import * as React from 'react'

  export interface ExcelFileProps {
    filename?: string;
    fileExtension?: string;
    element?: any; //Download Element
    children?: Array<React.ReactElement> | React.ReactElement; // Array<ExcelSheetProps>;
  }

  export interface ExcelSheetProps {
    name: string;
    data?: Array<object>;
    dataSet?: Array<ExcelSheetData>;
    value?: Array<string> | Function;
    children?: Array<React.ReactElement> | React.ReactElement; // Array<ExcelColumnProps>
  }

  export interface ExcelSheetData {
    xSteps?: number;
    ySteps?: number;
    columns: Array<string> | Array<ExcelCellHeader>;
    data: Array<Array<ExcelCellData>>;
  }

  export type ExcelCellData = ExcelValue | ExcelCell | Array<ExcelValue>;
  export type ExcelValue = string | number | Date | boolean;

  export interface ExcelCellHeader {
    title: string;
    style?: ExcelStyle;
  }

  export interface ExcelCell {
        value: ExcelValue;
        style?: ExcelStyle;
  }

  export interface ExcelColumnProps {
    label: string;
    value: number | boolean | string | Function;
  }

  export interface ExcelStyle {
    fill?: ExcelCellFillType;
    font?: ExcelFont;
    numFmt?: ExcelNumFormat;
    alignment?: ExcelAlignment;
    border?: ExcelBorder;
  }

  /* ExcelCell Fill Type */
  export type ExcelCellPatternType = "solid" | "none";

  export interface ExcelColorSpec {
    auto?: number; //default 1
    rgb?: string; //hex ARGB color
    theme?: ExcelTheme;
    indexed?: number;
  }

  export interface ExcelTheme {
    theme: string;
    tint: string;
  }

  export interface ExcelCellFillType {
    patternType?: ExcelCellPatternType;
    fgColor?: ExcelColorSpec;
    bgColor?: ExcelColorSpec;
  }

  /* Excel Font */
  export interface ExcelFont {
    name?: string;          // default `"Calibri"`
    sz?: number;             //font size in points default 11
    color?: ExcelColorSpec;
    bold?: boolean;
    underline?: boolean;
    italic?: boolean;
    strike?: boolean;
    outline?: boolean;
    shadow?: boolean;
    vertAlign?: boolean;
  }

  /* ExcelNumFormat */
  export type ExcelNumFormat = "0" | "0.00%" | "0.0%" | "0.00%;\\(0.00%\\);\\-;@" | "m/dd/yy" | string;

  /* ExcelAlignment */
  export interface ExcelAlignment {
    vertical?: ExcelAlignmentType;
    horizontal?: ExcelAlignmentType;
    wrapText?: boolean;
    readingOrder?: ExcelReadingOrder;
    textRotation?: ExcelTextRotation;
  }

  export type ExcelTextRotation = 0 | 45 | 90 | 135 | 180 | 255;

  export enum ExcelReadingOrder { LeftToRight = 1, RightToLeft}

  export type ExcelAlignmentType = "bottom" | "center" | "top";

  /* ExcelBorder */
  export interface ExcelBorder {
    style: ExcelBorderStyle;
    color: ExcelColorSpec;
  }

  export type ExcelBorderStyle =
    "thin"
    | "medium"
    | "thick"
    | "dotted"
    | "hair"
    | "dashed"
    | "mediumDashed"
    | "dashDot"
    | "mediumDashDot"
    | "dashDotDot"
    | "mediumDashDotDot"
    | "slantDashDot";

  export class ExcelColumn extends React.Component<ExcelColumnProps, any> {
  }

  export class ExcelSheet extends React.Component<ExcelSheetProps, any> {
  }

  export class ExcelFile extends React.Component<ExcelFileProps, any> {
  }

  export namespace ReactExport {
    export class ExcelFile extends React.Component<ExcelFileProps, any> {
      static ExcelSheet: React.ElementType<ExcelSheetProps>;
      static ExcelColumn: React.ElementType<ExcelColumnProps>;
    }
  }
  export default ReactExport
}
